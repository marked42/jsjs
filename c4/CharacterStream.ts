export class CharacterStream {
  static EOF = -1
  private index = 0
  constructor(private input: string) {}

  peek() {
    const code = this.input.charCodeAt(this.index)
    if (Number.isNaN(code)) {
      return CharacterStream.EOF
    }

    return code
  }

  next() {
    const code = this.peek()
    if (code !== CharacterStream.EOF) {
      this.index++
    }

    return code
  }

  prev() {
    this.index -= 1
  }
}
