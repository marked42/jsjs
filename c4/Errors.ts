export const StringLiteralEndEarly = new Error(
  'unexpected end of file when recognizing string token'
)
export const StringLiteralUnescapedNewline = new Error(
  'unescaped newline encountered when recognizing string token'
)

export const NumericLiteralUnexpectedDigitInOctalInteger = new Error(
  'unexpected digit 8 or 9 in octal digit, only 0-7 allowed'
)

export const NumericLiteralUnexpectedNonBinaryDigitAfter0B = new Error(
  'unexpected non binary digit after prefix 0b, expect at least one binary digit followed'
)

export const NumericLiteralUnexpectedNonHexDigitAfter0X = new Error(
  'unexpected non hex digit after prefix 0x, expect at least one hex digit followed'
)
