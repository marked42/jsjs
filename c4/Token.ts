export enum TokenType {
  EOF = -1,
  StringLiteral,
  Identifier,
  LineComment,
  NumericLiteral,
  Whitespace,
  Newline,

  // Keyword,
  Enum,
  Int,
  Char,
  If,
  Else,
  While,
  Return,

  // Operator
  Sizeof,
  Negate,
  Minus,
  MinusAssign,
  Plus,
  PlusAssign,
  Star,
  StarStar,
  StarAssign,
  Div,
  DivAssign,
  Tilde,
  And,
  AndAnd,
  Or,
  OrOr,
  Xor,
  Increment,
  Decrement,
  Modulus,
  ModulusAssign,
  Assign,
  Equal,
  NotEqual,
  Greater,
  GreaterEqual,
  Less,
  LessEqual,
  RightShift,
  LeftShift,

  // Punctuation
  LeftParen,
  RightParen,
  LeftSquare,
  RightSquare,
  LeftBracket,
  RightBracket,
  Question,
  Colon,
  Comma,
  SemiColon,
  Dot,
}

export const TokenEOF = {
  type: TokenType.EOF,
  source: '',
}

export interface Token {
  type: TokenType
  source: string
}

// TODO: refactor as class for simple construction
export interface StringLiteral extends Token {
  value: string
}

export interface Whitespace extends Token {
  value: string
}

export interface Newline extends Token {
  value: string
}

export interface Identifier extends Token {
  name: string
}

export interface LineComment extends Token {
  value: string
}

export interface NumericLiteral extends Token {
  value: number
}
