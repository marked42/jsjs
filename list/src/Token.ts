export class Token {
  constructor(public type: TokenType, public text: string) {}
}

export enum TokenType {
  Invalid = 'invalid',
  EOF = 'eof',
  Name = 'name',
  Comma = 'comma',
  Equal = 'equal',
  LeftBracket = 'leftBracket',
  RightBracket = 'rightBracket',
}
