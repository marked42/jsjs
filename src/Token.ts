export enum TokenType {
  WhiteSpace,
  LineTerminator,
  Comment,
  CommonToken,
  DivPunctuator,
  RightBracePunctuator,
  EOF,
}

export abstract class Token {
  constructor(protected _text: string) {}
  text() {
    return this._text
  }

  abstract type(): TokenType
}

export class WhiteSpace extends Token {
  type() {
    return TokenType.WhiteSpace
  }
}

// class LineTerminator extends Token {}

// class Comment extends Token {}

// class DivPunctuator extends Token {
//   constructor() {
//     super('/')
//   }
// }

// class RightBracePunctuator extends Token {
//   constructor() {
//     super('}')
//   }
// }

// class CommonToken extends Token {}

// class IdentifierName extends CommonToken {}

// class Punctuator extends CommonToken {}

// class NumericLiteral extends CommonToken {}

// class StringLiteral extends CommonToken {}

enum GoalSymbol {
  InputElementDiv,
  InputElementRegExp,
  InputElementTemplateTail,
  InputElementRegExpOrTemplateTail,
}
