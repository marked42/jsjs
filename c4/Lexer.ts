import { CharacterStream } from './CharacterStream'
import { Token, TokenType, Identifier, StringLiteral } from './Token'
import * as Errors from './Errors'

export class Lexer {
  private codes: number[] = []
  private charCode: number = -1
  constructor(private stream: CharacterStream) {}

  getChar() {
    this.charCode = this.stream.next()
    if (this.charCode !== CharacterStream.EOF) {
      this.codes.push(this.charCode)
    }
    return this.charCode
  }

  ungetChar() {
    this.codes.splice(this.codes.length - 1, 1)
    this.stream.prev()
    this.charCode = this.codes[this.codes.length - 1]
  }

  resetCharCodes() {
    this.codes = []
    this.charCode = -1
  }

  next(): Token {
    this.getChar()
    // StringLiteral
    if (charCodeIs(this.charCode, '"')) {
      const value = []

      while (true) {
        this.getChar()

        if (charCodeIs(this.charCode, '\\')) {
          this.getChar()
          if (charCodeIs(this.charCode, '"')) {
            value.push('"'.charCodeAt(0))
          } else if (charCodeIs(this.charCode, "'")) {
            value.push("'".charCodeAt(0))
          } else if (charCodeIs(this.charCode, 'b')) {
            value.push('\b'.charCodeAt(0))
          } else if (charCodeIs(this.charCode, 'f')) {
            value.push('\f'.charCodeAt(0))
          } else if (charCodeIs(this.charCode, 'n')) {
            value.push('\n'.charCodeAt(0))
          } else if (charCodeIs(this.charCode, 'r')) {
            value.push('\r'.charCodeAt(0))
          } else if (charCodeIs(this.charCode, 't')) {
            value.push('\t'.charCodeAt(0))
          } else if (charCodeIs(this.charCode, 'v')) {
            value.push('\v'.charCodeAt(0))
          } else if (charCodeIs(this.charCode, '\\')) {
            value.push('\\'.charCodeAt(0))
          } else if (charCodeIs(this.charCode, '\r')) {
            value.push('\r'.charCodeAt(0))
            this.getChar()
            if (charCodeIs(this.charCode, '\n')) {
              value.push('\n'.charCodeAt(0))
            } else {
              this.ungetChar()
            }
          } else if (charCodeIs(this.charCode, '\n')) {
            value.push('\n'.charCodeAt(0))
          } else {
            console.error(
              'invalid escape sequence \\' + String.fromCharCode(this.charCode)
            )
            value.push(this.charCode)
          }
        } else if (
          charCodeIs(this.charCode, '\r') ||
          charCodeIs(this.charCode, '\n')
        ) {
          throw Errors.StringLiteralUnescapedNewline
        } else if (charCodeIs(this.charCode, '"')) {
          const source = String.fromCharCode(...this.codes)
          this.resetCharCodes()
          return {
            type: TokenType.StringLiteral,
            source,
            value: String.fromCharCode(...value),
          } as StringLiteral
        } else if (this.charCode === CharacterStream.EOF) {
          throw Errors.StringLiteralEndEarly
        } else {
          value.push(this.charCode)
        }
      }
      // NumericLiteral
    } else if (isDigit(this.charCode)) {
      // LineComment
    } else if (charCodeIs(this.charCode, '/')) {
      this.stream.next()

      if (charCodeIs(this.stream.peek(), '/')) {
        this.stream.next()
        const codes = []
        while (true) {
          const code = this.stream.peek()
          if (!charCodeIs(code, '\r') && !charCodeIs(code, '\n')) {
            codes.push(code)
            this.stream.next()
          } else {
            return {
              type: TokenType.LineComment,
              source: String.fromCharCode(...codes),
            }
          }
        }
      } else {
        return {
          type: TokenType.Operator,
          source: '/',
        }
      }
      // Whitespace
    } else if (isWhitespace(this.charCode)) {
      const codes = [this.stream.next()]
      while (isWhitespace(this.stream.peek())) {
        codes.push(this.stream.next())
      }

      return {
        type: TokenType.Whitespace,
        source: String.fromCharCode(...codes),
      }
      // Newline
    } else if (charCodeIs(this.charCode, '\r')) {
      this.stream.next()
      if (charCodeIs(this.stream.peek(), '\n')) {
        this.stream.next()
        return {
          type: TokenType.Newline,
          source: '\r\n',
        }
      } else {
        return {
          type: TokenType.Newline,
          source: '\r',
        }
      }
    } else if (charCodeIs(this.charCode, '\n')) {
      this.stream.next()
      return {
        type: TokenType.Whitespace,
        source: '\n',
      }
    } else if (isIdentifierStart(this.charCode)) {
      const codes = [this.stream.next()]

      while (true) {
        const code = this.stream.peek()
        if (isIdentifierContinue(code)) {
          codes.push(code)
          this.stream.next()
          continue
        }

        const source = String.fromCharCode(...codes)
        return {
          type: TokenType.Identifier,
          source,
          name: source,
        } as Identifier
      }
    }

    throw new Error('Unexpected token ' + String.fromCharCode(this.charCode))
  }
}

function charCodeIs(code: number, text: string) {
  if (text.length !== 1) {
    throw new Error('text must be single character, got ' + text)
  }
  return code === text.charCodeAt(0)
}

function isDigit(code: number) {
  return code >= '0'.charCodeAt(0) && code <= '9'.charCodeAt(0)
}

function isWhitespace(code: number) {
  const ws = [' ', '\t'].map((c) => c.charCodeAt(0))

  return ws.includes(code)
}

function isLetter(code: number) {
  const isLower = code >= 'a'.charCodeAt(0) && code <= 'z'.charCodeAt(0)
  const isUpper = code >= 'A'.charCodeAt(0) && code <= 'Z'.charCodeAt(0)

  return isLower || isUpper
}

function isIdentifierStart(code: number) {
  return isLetter(code) || charCodeIs(code, '_')
}

function isIdentifierContinue(code: number) {
  return isLetter(code) || charCodeIs(code, '_') || isDigit(code)
}
