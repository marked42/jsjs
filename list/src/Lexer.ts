import { Token } from './Token'

export abstract class Lexer {
  public index = 0
  public char = Lexer.EOF

  static EOF = -1

  constructor(protected input: string) {
    this.char = input.charCodeAt(0)
  }

  abstract nextToken(): Token

  consume() {
    this.index++
    if (this.index >= this.input.length) {
      this.char = Lexer.EOF
    } else {
      this.char = this.input.charCodeAt(this.index)
    }
  }

  match(char: number) {
    if (this.char !== char) {
      throw new Error(`expecting ${char}, get ${this.char}`)
    }
    this.consume()
  }
}
