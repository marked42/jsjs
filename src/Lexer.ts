import { Token, TokenType, WhiteSpace } from './Token'

type CodePoint = number
const EOF = 12

abstract class CodePointStream {
  abstract peek(): CodePoint
  abstract next(): CodePoint
  end() {
    return this.peek() === EOF
  }
}

abstract class TokenStream {
  abstract peek(): Token
  abstract next(): Token
  end() {
    return this.peek().type() === TokenType.EOF
  }
}

class Lexer {
  constructor(private codePointStream: CodePointStream) {}

  next() {}

  tokens(): TokenStream {
    return {
      peek() {
        return new WhiteSpace('')
      },
      next() {
        return new WhiteSpace('')
      },
      end() {
        return false
      },
    }
  }
}
