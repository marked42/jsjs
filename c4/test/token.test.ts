import { Lexer } from '../Lexer'
import { CharacterStream } from '../CharacterStream'
import { TokenEOF, TokenType, Token } from '../Token'
import * as Errors from '../Errors'

function recognizeToken(input: string) {
  const stream = new CharacterStream(input)
  const lexer = new Lexer(stream)

  return lexer._next()
}

describe('StringLiteral', () => {
  it.each([
    [
      'recognize normal string literal',
      '"test"',
      {
        type: TokenType.StringLiteral,
        source: '"test"',
        value: 'test',
      },
    ],
    [
      'recognize escape sequence newline',
      '"\\n"',
      {
        type: TokenType.StringLiteral,
        source: '"\\n"',
        value: '\n',
      },
    ],
    [
      'recognize escape sequence backward',
      '"\\b"',
      {
        type: TokenType.StringLiteral,
        source: '"\\b"',
        value: '\b',
      },
    ],
    [
      'recognize escape sequence formfeed',
      '"\\f"',
      {
        type: TokenType.StringLiteral,
        source: '"\\f"',
        value: '\f',
      },
    ],
    [
      'recognize escape sequence carriage return',
      '"\\r"',
      {
        type: TokenType.StringLiteral,
        source: '"\\r"',
        value: '\r',
      },
    ],
    [
      'recognize escape sequence tab',
      '"\\t"',
      {
        type: TokenType.StringLiteral,
        source: '"\\t"',
        value: '\t',
      },
    ],
    [
      'recognize escape sequence vertical tab',
      '"\\v"',
      {
        type: TokenType.StringLiteral,
        source: '"\\v"',
        value: '\v',
      },
    ],
    [
      'recognize escape sequence double quote',
      '"\\""',
      {
        type: TokenType.StringLiteral,
        source: '"\\""',
        value: '"',
      },
    ],
    [
      'recognize escape sequence single quote',
      '"\\\'"',
      {
        type: TokenType.StringLiteral,
        source: '"\\\'"',
        value: "'",
      },
    ],
    [
      'recognize escape sequence backward slash',
      '"\\\\"',
      {
        type: TokenType.StringLiteral,
        source: '"\\\\"',
        value: '\\',
      },
    ],
    [
      'recognize escape sequence line continuation',
      '"\\\n"',
      {
        type: TokenType.StringLiteral,
        source: '"\\\n"',
        value: '\n',
      },
    ],
    [
      'recognize escape sequence line continuation',
      '"\\\r"',
      {
        type: TokenType.StringLiteral,
        source: '"\\\r"',
        value: '\r',
      },
    ],
    [
      'recognize escape sequence line continuation',
      '"\\\r\n"',
      {
        type: TokenType.StringLiteral,
        source: '"\\\r\n"',
        value: '\r\n',
      },
    ],
    [
      'throw error when encountering early EOF',
      '"',
      Errors.StringLiteralEndEarly,
    ],
    [
      'throw error when encountering unescaped newline',
      '"\n"',
      Errors.StringLiteralUnescapedNewline,
    ],
  ])('%s', (_title, input, result) => {
    if (result instanceof Error) {
      expect(() => recognizeToken(input)).toThrowError(result)
    } else {
      expect(recognizeToken(input)).toEqual(result)
    }
  })
})

describe('LineComment', () => {
  it('recognize line comment with content', () => {
    const input = '// line1\rline2'
    expect(recognizeToken(input)).toEqual({
      type: TokenType.LineComment,
      source: '// line1',
      value: ' line1',
    })
  })

  it('recognize line comment with content at the end of file', () => {
    const input = '// line1'
    expect(recognizeToken(input)).toEqual({
      type: TokenType.LineComment,
      source: '// line1',
      value: ' line1',
    })
  })

  it('recognize line comment with empty content immediately followed by newline', () => {
    const inputs = ['//\n', '//\r', '//\r\n']
    inputs.forEach((input) => {
      expect(recognizeToken(input)).toEqual({
        type: TokenType.LineComment,
        source: '//',
        value: '',
      })
    })
  })

  it('recognize line comment with empty content at the end of file', () => {
    const input = '//'
    expect(recognizeToken(input)).toEqual({
      type: TokenType.LineComment,
      source: '//',
      value: '',
    })
  })
})

describe('Whitespace', () => {
  it('recognize whitespace', () => {
    const input = ' '
    expect(recognizeToken(input)).toEqual({
      type: TokenType.Whitespace,
      source: ' ',
    })
  })

  it('recognize whitespace tab', () => {
    const input = '\t'
    expect(recognizeToken(input)).toEqual({
      type: TokenType.Whitespace,
      source: '\t',
    })
  })

  it('recognize whitespace of multiple characters', () => {
    const input = ' \t \t'
    expect(recognizeToken(input)).toEqual({
      type: TokenType.Whitespace,
      source: ' \t \t',
    })
  })
})

describe('Newline', () => {
  it('recognize \\r\\n as newline', () => {
    const input = '\r\n'
    expect(recognizeToken(input)).toEqual({
      type: TokenType.Newline,
      source: '\r\n',
    })
  })

  it('recognize \\r as newline', () => {
    const input = '\r'
    expect(recognizeToken(input)).toEqual({
      type: TokenType.Newline,
      source: '\r',
    })
  })

  it('recognize \\n as newline', () => {
    const input = '\n'
    expect(recognizeToken(input)).toEqual({
      type: TokenType.Newline,
      source: '\n',
    })
  })
})

describe('Identifier', () => {
  it.each([
    [
      'a',
      {
        type: TokenType.Identifier,
        source: 'a',
        name: 'a',
      },
    ],
    [
      'A',
      {
        type: TokenType.Identifier,
        source: 'A',
        name: 'A',
      },
    ],
    [
      '_',
      {
        type: TokenType.Identifier,
        source: '_',
        name: '_',
      },
    ],
    [
      '_a1',
      {
        type: TokenType.Identifier,
        source: '_a1',
        name: '_a1',
      },
    ],
    [
      'aa1',
      {
        type: TokenType.Identifier,
        source: 'aa1',
        name: 'aa1',
      },
    ],
    [
      'AA1',
      {
        type: TokenType.Identifier,
        source: 'AA1',
        name: 'AA1',
      },
    ],
  ])('recognize identifier', (input, result) => {
    expect(recognizeToken(input)).toEqual(result)
  })
})

describe('NumericLiteral', () => {
  it.each([
    [
      'single digit zero',
      '0',
      {
        type: TokenType.NumericLiteral,
        source: '0',
        value: 0,
      },
    ],
    [
      'non-zero single digit',
      '1',
      {
        type: TokenType.NumericLiteral,
        source: '1',
        value: 1,
      },
    ],
    [
      'with multiple decimal digits followed by non digit',
      '121as',
      {
        type: TokenType.NumericLiteral,
        source: '121',
        value: 121,
      },
    ],
    [
      'with multiple decimal digits followed by EOF',
      '121',
      {
        type: TokenType.NumericLiteral,
        source: '121',
        value: 121,
      },
    ],
    [
      'with binary digits',
      '0b11',
      {
        type: TokenType.NumericLiteral,
        source: '0b11',
        value: 3,
      },
    ],
    [
      'throw error with no following binary digit',
      '0b',
      Errors.NumericLiteralUnexpectedNonBinaryDigitAfter0B,
    ],
    [
      'with octal digits',
      '011',
      {
        type: TokenType.NumericLiteral,
        source: '011',
        value: 9,
      },
    ],
    [
      'throw error with no following octal digit',
      '08',
      Errors.NumericLiteralUnexpectedDigitInOctalInteger,
    ],
    [
      'hex digits',
      '0x11',
      {
        type: TokenType.NumericLiteral,
        source: '0x11',
        value: 17,
      },
    ],
    [
      'hex digits',
      '0x0a',
      {
        type: TokenType.NumericLiteral,
        source: '0x0a',
        value: 10,
      },
    ],
    [
      'throw error with no following hex digit',
      '0xg',
      Errors.NumericLiteralUnexpectedNonHexDigitAfter0X,
    ],
  ])('recognize numeric literal %s', (_title, input, result) => {
    if (result instanceof Error) {
      expect(() => recognizeToken(input)).toThrowError(result)
    } else {
      expect(recognizeToken(input)).toEqual(result)
    }
  })
})

describe('Keyword', () => {
  it('recognize keyword', () => {
    const keywords = [
      ['enum', TokenType.Enum],
      ['int', TokenType.Int],
      ['char', TokenType.Char],
      ['if', TokenType.If],
      ['else', TokenType.Else],
      ['return', TokenType.Return],
      ['while', TokenType.While],
    ] as const

    keywords.forEach((option) => {
      const [keyword, type] = option

      expect(recognizeToken(keyword)).toEqual({
        type,
        source: keyword,
      })
    })
  })
})

describe('Operator and Punctuation', () => {
  it('recognize operator', () => {
    const keywords = [
      ['sizeof', TokenType.Sizeof],
      ['+', TokenType.Plus],
      ['++', TokenType.Increment],
      ['+=', TokenType.PlusAssign],
      ['-', TokenType.Minus],
      ['--', TokenType.Decrement],
      ['-=', TokenType.MinusAssign],
      ['*', TokenType.Star],
      ['**', TokenType.StarStar],
      ['*=', TokenType.StarAssign],
      ['/', TokenType.Div],
      ['/=', TokenType.DivAssign],
      ['%', TokenType.Modulus],
      ['%=', TokenType.ModulusAssign],
      ['~', TokenType.Tilde],
      ['&', TokenType.And],
      ['&=', TokenType.AndAssign],
      ['&&', TokenType.AndAnd],
      ['|', TokenType.Or],
      ['|=', TokenType.OrAssign],
      ['||', TokenType.OrOr],
      ['^', TokenType.Xor],
      ['^=', TokenType.XorAssign],
      ['=', TokenType.Assign],
      ['==', TokenType.Equal],
      ['!', TokenType.Negate],
      ['!=', TokenType.NotEqual],
      ['>', TokenType.Greater],
      ['>=', TokenType.GreaterEqual],
      ['<', TokenType.Less],
      ['<=', TokenType.LessEqual],
      ['>>', TokenType.RightShift],
      ['>>=', TokenType.RightShiftAssign],
      ['<<', TokenType.LeftShift],
      ['<<=', TokenType.LeftShiftAssign],

      ['(', TokenType.LeftParen],
      [')', TokenType.RightParen],
      ['{', TokenType.LeftBracket],
      ['}', TokenType.RightBracket],
      ['[', TokenType.LeftSquare],
      [']', TokenType.RightSquare],
      ['?', TokenType.Question],
      [':', TokenType.Colon],
      [',', TokenType.Comma],
      [';', TokenType.SemiColon],
      ['.', TokenType.Dot],
      ['->', TokenType.PointerMember],
    ] as const

    keywords.forEach((option) => {
      const [keyword, type] = option

      expect(recognizeToken(keyword)).toEqual({
        type,
        source: keyword,
      })
    })
  })
})

describe('multiple tokens of different types', () => {
  it('should recognize different types of tokens', () => {
    const input = `
      int a = 0b01;
      // line comment
      char* b = "test";

      int c = a == b;
    `

    const stream = new CharacterStream(input)
    const lexer = new Lexer(stream)

    const tokens: Token[] = []
    while (true) {
      const token = lexer.consume()
      tokens.push(token)
      if (token === TokenEOF) {
        break
      }
    }

    expect(tokens).toMatchSnapshot()
  })
})
