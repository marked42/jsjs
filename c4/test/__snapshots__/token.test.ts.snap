// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`multiple tokens of different types should recognize different types of tokens 1`] = `
Array [
  Object {
    "source": "
",
    "type": 5,
  },
  Object {
    "source": "      ",
    "type": 4,
  },
  Object {
    "source": "int",
    "type": 7,
  },
  Object {
    "source": " ",
    "type": 4,
  },
  Object {
    "name": "a",
    "source": "a",
    "type": 1,
  },
  Object {
    "source": " ",
    "type": 4,
  },
  Object {
    "source": "=",
    "type": 34,
  },
  Object {
    "source": " ",
    "type": 4,
  },
  Object {
    "source": "0b01",
    "type": 3,
    "value": 1,
  },
  Object {
    "source": ";",
    "type": 52,
  },
  Object {
    "source": "
",
    "type": 5,
  },
  Object {
    "source": "      ",
    "type": 4,
  },
  Object {
    "source": "// line comment",
    "type": 2,
    "value": " line comment",
  },
  Object {
    "source": "
",
    "type": 5,
  },
  Object {
    "source": "      ",
    "type": 4,
  },
  Object {
    "source": "char",
    "type": 8,
  },
  Object {
    "source": "*",
    "type": 19,
  },
  Object {
    "source": " ",
    "type": 4,
  },
  Object {
    "name": "b",
    "source": "b",
    "type": 1,
  },
  Object {
    "source": " ",
    "type": 4,
  },
  Object {
    "source": "=",
    "type": 34,
  },
  Object {
    "source": " ",
    "type": 4,
  },
  Object {
    "source": "\\"test\\"",
    "type": 0,
    "value": "test",
  },
  Object {
    "source": ";",
    "type": 52,
  },
  Object {
    "source": "
",
    "type": 5,
  },
  Object {
    "source": "
",
    "type": 5,
  },
  Object {
    "source": "      ",
    "type": 4,
  },
  Object {
    "source": "int",
    "type": 7,
  },
  Object {
    "source": " ",
    "type": 4,
  },
  Object {
    "name": "c",
    "source": "c",
    "type": 1,
  },
  Object {
    "source": " ",
    "type": 4,
  },
  Object {
    "source": "=",
    "type": 34,
  },
  Object {
    "source": " ",
    "type": 4,
  },
  Object {
    "name": "a",
    "source": "a",
    "type": 1,
  },
  Object {
    "source": " ",
    "type": 4,
  },
  Object {
    "source": "==",
    "type": 35,
  },
  Object {
    "source": " ",
    "type": 4,
  },
  Object {
    "name": "b",
    "source": "b",
    "type": 1,
  },
  Object {
    "source": ";",
    "type": 52,
  },
  Object {
    "source": "
",
    "type": 5,
  },
  Object {
    "source": "    ",
    "type": 4,
  },
  Object {
    "type": -1,
  },
]
`;
